Code review for Fix#11335

ReggieDoodle opened the pull request:
This is my first exposure to the UI in cgeo - well any android app actually - so I would like some feedback please. I’m not expecting it to be ‘production ready’.

Ticket #11335 outlines some issues with the current coordinate entry dialog and this is my attempt to recreate the same functionality in an alert based dialog. 

The design code (xml) is quite verbose atm but that was how it grew as I learned what various options did.  It can be styled and compressed no doubt.

The dialog logic is similar to the existing dialogs - some code was lifted straight from it - this is labelled as such. The main difference is that the existing one uses binding, though not in the sense I am used to it (WPF view models) so I haven’t recreated that approach.

My current approach returns a string representing latitude and longitude via the callback - this maybe should just return the validated geopoint to save parsing it twice - but it ended up like that as the origanal UI just took a string anyway so it mimicked that.

The UI in the distance filter itself has a new button to launch the dialog, as per usages elsewhere but the slider representing distance hasn’t been touched. It could so with some improvement, as per my comment on the ticket

Thanks
Dave

ReggieDoodle commented:
Codacy issues address and branch updated

ReggieDoodle commented:
Fixed two more that were masked by previous

ReggieDoodle commented:
All tests pass - can someone review the code please

moving-bits commented:
ok to test

ReggieDoodle commented:
Resynced my branch as there appears to have been some CI changes - check secrets ??

ReggieDoodle commented:
@kumy - getting permission issues when the unit tests run following a commit - seems you addressed this yesterday (https://github.com/cgeo/cgeo/issues/16804).   I've resynched my fork on master. Anything else I need to do please?

kumy commented:
Hi @ReggieDoodle, thanks for the ping. I did other changes today, you may need to rebase again.

Then as you have "read" privilege (ie not a team member) over the cgeo repo, you need your PR to be first reviewed by one team member (/cc @Lineflyer @moving-bits) and at their discretion they should add the label "safe to test".

kumy commented:
I did a quick PR content check and will add the label myself.

ReggieDoodle commented:
Ok thanks. Not sure what its telling me as it's a branch in my repo where I obviously have full access ?

kumy commented:
A branch from your fork, running on cgeo side as a PR. :)
As non cgeo/cgeo repo member we need to allow your PR to run (security measure) by adding the label "safe to test".

ReggieDoodle commented:
> A branch from your fork, running on cgeo side as a PR. :)
> As non cgeo/cgeo repo member we need to allow your PR to run (security measure) by adding the label "safe to test".

Is this a new thing as my PRs have run ok before - normally one of the core team have just added a comment saying ok to test - not seen the label before ?

kumy commented:
Yes I readded the Unit Test on GitHub side a few days ago.

ReggieDoodle commented:
Rebased on current master as branch was a week old 

Added sub branch Fix#11335-A to call my new dialog in place of existing dialog to prove it can be used

The old dialog is used by these functions

1. User defined cache set coordinates - easy change
2. Waypoint editor - quite complex
3. Search filter
4. Logging a trackable - not sure why that needs coords atm

Skipping the CCG  dialog for the time being

moving-bits commented:
Sorry for not coming back to you earlier - everyone in the team (including me) seems to be quite busy currently.

> Rebased on current master as branch was a week old

FYI: Generally there is no need for it, as long as there are no conflicting changes on the branch your PR is targeted to - GitHub will rebase your PR automatically on merging.

> 4. Logging a trackable - not sure why that needs coords atm

IIRC, the GeoKrety type of trackables requires coordinates on logging, so you should not see this dialog when logging a geocaching.com trackable, but it should be visible when logging a GeoKrety.

> Skipping the CCG dialog for the time being

What is "CCG"?

With regard to a code review: We have a core team virtual meeting today, I added this item to the agenda. Hopefully someone finds the time for a more thorough look at your PR.

Anyway, thanks for diving into this complex thing!

P. S.: I have changed the PR's title to a more descriptive title to make it easier to understand at a first glance what this PR is about. Hope that's ok for you.

ReggieDoodle commented:
Thanks for the comment. Nice to hear what you guys think so far.

CCG is the calculate coordinates global dialog - the ok that lets you compute a waypoint using A B C etc.    just skipped if for the time being. Might also need to be converted or just called as is but it's another fragment based one

ReggieDoodle commented:
GeoKrety Trackable logging updated and tested in branch Fix#11335-A.

This required passing in a reference to a cache as well as a coordinate so added a new show method which will also be needed when I address the waypoint usage

ReggieDoodle commented:
I've taken this to a suitable stage that, once reviewed, I believe it can be put into use.  The distance filter, user defined caches and Geokrety TB pages can all use the new dialog now.

The next stage is to address waypoints but as said previously the calculate coordinates global dialog (CCG) is tightly coupled to the existing coordinate entry dialog and it also uses a fragment manager so it can't be called from the new alert dialog as it stands.  Before diving in and converting that (it's a complex beast) I would like to refactor the waypoint page a little so there is a common dialog that calls either the coordinate entry dialog or the CCG rather than one call the other as it is now - it could probably include the projection options too? 

That would all need quite a bit of planning and some user stories / mock-ups creating.

In the meantime I believe the old and new approach can live side by side

Dave

murggel commented:
I  will have a look at the functionality and handling of the new dialog, but I will only find time after the easter weekend.

ReggieDoodle commented:
@murggel - thanks for offering to review this 

I've updated the branch this morning - it now uses binding and I have fixed a logic issue in the distance filter

The dialog is **only** used by the distance filter in this branch. I have another branch where it is used by the search card, user defined cache and GeoKrety TB pages which I was going to submit once this was reviewed (unless you want it all in one ?)

eddiemuc commented on main/src/main/java/cgeo/geocaching/filters/gui/DistanceFilterViewHolder.java:
```diff
@@ -38,10 +40,20 @@ public View createView() {
+        useCurrentPosition.setOnClickListener(v -> toggleCurrent());
+        location = LocationDataProvider.getInstance().currentGeo().getCoords();
+
+        setCoordsButton = ViewUtils.createButton(getActivity(), ll, TextParam.id(R.string.cache_filter_distance_coordinates));
```
Format of the coord button should not be set explicitely in code. 

There's a style already defined for displaying a coordinate button: @style/button_full_double
It is used eg in editwaypoint_activity.xml

To use it here, you may use the variant of ViewUtils.creatButton which allows passing a layout file. You need to create such a layout file first. An example for this is button_wherigo_view.xml (copy it and replace the used style inside)

ReggieDoodle replied:
I mentioned styling in the intro. I wasn’t sure about using them and tbh I wanted to code each bit to see what / how they worked since it’s my first go at this but can look at incorporating some now it’s working

ReggieDoodle replied:
Done

eddiemuc commented on main/src/main/java/cgeo/geocaching/ui/dialog/DialogCallback.java:
```diff
@@ -0,0 +1,5 @@
+package cgeo.geocaching.ui.dialog;
+
+public interface DialogCallback {
+    void onDialogClosed(String input);
```
* Why not passing a Geopoint object instead of a string?
* You may want to get rid of class `DialogCallback` completely. Instead you can use `Consumer<Geopoint>`

ReggieDoodle replied:
There is a note at the top of the page why it’s a string but now I can change that now the UI is working.

Not familiar with the consumer<T> - will look at it though. Is there an example somewhere in the existing code please ?

ReggieDoodle replied:
Callback changed to take a Geopoint.

eddiemuc commented on main/src/main/java/cgeo/geocaching/ui/dialog/NewCoordinateInputDialog.java:
```diff
@@ -0,0 +1,438 @@
+    private List<EditText> orderedInputs;
+    private Geopoint gp, cacheCoords;
+
+    public NewCoordinateInputDialog(final Context context, final DialogCallback callback) {
```
You may want to make the constructor private and offer static methods instead which show a dialog in one line:
* `public static void show(final Context context, final Consumer<Geopoint> callback, final Geopoint cacheCoordinate)`
* `public static void show(final Context context, final Consumer<Geopoint> callback)`

eddiemuc replied:
Then leave as is. It's also fine

ReggieDoodle replied:
@eddiemuc - had another look - is this the correct approach for using a static show please?

![image](https://github.com/user-attachments/assets/16e73fb8-686a-4ad1-b570-f6fbceb6cce2)

ReggieDoodle replied:
Ok, it effects the other usages of the dialog (eg user defined cache) so I'll do it in that branch and make a new PR when this one is merged else this one will be going on forever !

eddiemuc commented on main/src/main/java/cgeo/geocaching/ui/dialog/NewCoordinateInputDialog.java:
```diff
@@ -0,0 +1,438 @@
+        builder.setView(theView);
+
+        final AlertDialog dialog = builder.create();
+        dialog.show();
```
`dialog.show()` should be called AFTER the GUI is initialized, not before

ReggieDoodle replied:
Done

eddiemuc commented on main/src/main/res/drawable/border_active.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
That one is new is it?

ReggieDoodle replied:
Yes. I wasn’t sure how to get an orange border so I made a new one.

eddiemuc reviewed (approved):
LGTM! Thanks for all the work!

I left some comments, but there is no blocker for merge among them.

murggel commented:
I found some time today to look into the UI.

LGTM - thanks for the effort.

At least I have encountered one bug and one issue, which probably will be solved with using the suggestions of @eddiemuc 

And some minor UI-glinches

I will add one comment per observation for better tracking

murggel commented:
Bug with format DDD.DDDD° - "c:geo can't parse latitude or longitude"
![image](https://github.com/user-attachments/assets/a7a0dc5e-5201-4529-9964-c73937e03980)

murggel commented:
What we should discuss is the "default"-format for the plain-format
Normally we have the DDD° MM.MMM' format, but here it is the DDD.DDDDD° format

| old dialog | new dialog |
| -------- | --------- |
|![image](https://github.com/user-attachments/assets/9ddd5cc5-2437-415e-aba3-62201b28d93b)|![image](https://github.com/user-attachments/assets/a8154505-34f1-40e8-8c23-deba21b448a4)|

murggel commented:
Minor UI-issues:
It would be nice, if the ° and ' signs are in the input fields of the corresponding format-fields
![image](https://github.com/user-attachments/assets/009fb32e-d280-4279-a624-afb17633740c)

murggel commented:
Minor UI-issue:

Add captions to the input field as in the old dialog

![image](https://github.com/user-attachments/assets/57313a02-3770-4c95-94c7-e7dd0c739ee0)

murggel commented:
Minor UI-issue:

Add possibility to delete content of the field

![image](https://github.com/user-attachments/assets/2255052a-599a-4abd-8344-583ed6e420d5)

murggel commented:
To discuss for consistency:

in old dialog the button "From clipboard" is only shown, if there are coordinates in the clipboard

ReggieDoodle commented:
@murggel - thank you for looking at this.  I will address the issues after the weekend - out caching tmrw and family here for a big meal Sunday
Dave

ReggieDoodle commented:
@eddiemuc 

Quick look at the bug reported by @murggel when using DD.DDDD coordinate format

The string I'm passing in is N53.64340 W001.73898 which I would expect it to parse okay but seems it doesn't like the N & W and expects 53.64340 -001.73898 instead.

Easy to fix in my UI but I'd say that's a bug in the parser, not my use of it ?

ReggieDoodle commented:
> Bug with format DDD.DDDD° - "c:geo can't parse latitude or longitude" 

Fixed - I assumed the geo parser would accept N53.4445 W01.2585 but it doesn't like the hemisphere fields

ReggieDoodle commented:
> Minor UI-issue:
> 
> Add captions to the input field as in the old dialog

Fixed - Reused layout from existing dialog

ReggieDoodle commented:
> Minor UI-issue:
> 
> Add possibility to delete content of the field

Fixed with prebiosu issue

ReggieDoodle commented:
@eddiemuc & @murggel 

I have fixed most of the issues and commented on the outstanding ones -  squashed my changes and pushed to the original branch  

Still need to modify the second branch which implements the search card, user defined cache and GK TB page to use this dialog but rather get this first one merged please before adding those

moving-bits commented:
Thanks, @ReggieDoodle, for your extensive work, very welcomed!

Build system has reported two minor lint issues:
- "Unused import - cgeo.geocaching.location.GeopointParser" in `DistanceFilterViewHolder.java`
- "Wrong order for 'java.lang.ref.WeakReference' import" in `NewCoordinateInputDialog.java` (configured order is cgeo, android, androix, java (separate blocks each), then all others (common block))

ReggieDoodle commented:
Tks. I left it running and was going to check it after lunch. 

How do I run these checks locally before pushing ?  I still can't get unit test to run locally though

moving-bits commented:
> I still can't get unit test to run locally though

For the unit tests you'll need to configure all the keys in `keys.xml` file (see [API keys section in our readme file](https://github.com/cgeo/cgeo/blob/master/README.md#api-keys) and use a geocaching.com premium account configured in c:geo. Most of the keys can be obtained by yourself, only the key for adventure labs API access is bound to c:geo org and cannot be made public, so those tests will fail.

ReggieDoodle commented:
> > I still can't get unit test to run locally though
> 
> For the unit tests you'll need to configure all the keys in `keys.xml` file (see [API keys section in our readme file](https://github.com/cgeo/cgeo/blob/master/README.md#api-keys) and use a geocaching.com premium account configured in c:geo. Most of the keys can be obtained by yourself, only the key for adventure labs API access is bound to c:geo org and cannot be made public, so those tests will fail.

Its not that they fail - they don't run at all - I'll make a separate discussion ticket for it

ReggieDoodle commented:
> > How do I run these checks locally before pushing ? I still can't get unit test to run locally though
> 
> Most of the "typical" things will be found by running "CheckStyle", which you can install as a plugin to Android Studio, if it isn't installed already.

Can't find that anywhere !   I'm on AS Labybug

moving-bits commented:
> > > How do I run these checks locally before pushing ? I still can't get unit test to run locally though
> > 
> > 
> > Most of the "typical" things will be found by running "CheckStyle", which you can install as a plugin to Android Studio, if it isn't installed already.
> 
> Can't find that anywhere ! I'm on AS Labybug

I'm on Meerkat already, but the entry point should be the same: Go to File - Settings menu, there is a "plugins" section in which you can manage the installed plugins:

![image](https://github.com/user-attachments/assets/44c6e691-277a-4123-b7e5-68a4195bf3e7)

ReggieDoodle commented:
> I'm on Meerkat already, but the entry point should be the same: Go to File - Settings menu, there is a "plugins" section in which you can manage the installed plugins:

Thanks - I was looking for a internal one, not a third party - it loads but errors when running - I'll have a play

moving-bits commented:
> it loads but errors when running - I'll have a play

Does it "find errors in the source when running" or "does not run due to errors in plugin/AS"?

If the first, please select the "cgeo" profile in CheckStyle first.

If the latter: It might be that the CheckStyle version you have installed is not compatible to the AS version you use, see https://plugins.jetbrains.com/plugin/1065-checkstyle-idea/versions/stable

ReggieDoodle commented:
Ok, I'm done. ready to merge ?

moving-bits commented:
> I get this - does it mean the windows event viewer (nothing there) osr is there a separate log file somewhere?

This looks similar what I got some time ago: #13871. Recommended action back then was downgrading CheckStyle.

Not sure about the event log location, though, - probably in the LogCat window of AS?

bekuno commented:
You can run in the command window './gradlew check'

murggel commented:
> @eddiemuc & @murggel 
> 
> I have fixed most of the issues and commented on the outstanding ones -  squashed my changes and pushed to the original branch  
> 

I will have a look at it tomorrow evening...

murggel commented:
> > Minor UI-issue:
> > Add captions to the input field as in the old dialog
> 
> Fixed - Reused layout from existing dialog

Still found some minor issues with layout:

In the original layout the hint for the min / sec / fractions are set in the code-file, not the layout-file, that's why they are here missing. 

| original dialog | new dialog |
| -- | -- |
|![image](https://github.com/user-attachments/assets/b1544a4e-cc1d-4836-a49a-743a9940ad7f) | ![image](https://github.com/user-attachments/assets/0422553b-79c2-4c5a-a330-446a7ecf06c6)|

ReggieDoodle commented:
> > > Minor UI-issue:
> > > Add captions to the input field as in the old dialog
> > 
> > 
> > Fixed - Reused layout from existing dialog
> 
> Still found some minor issues with layout:
> 
> In the original layout the hint for the min / sec / fractions are set in the code-file, not the layout-file, that's why they are here missing.
> 
Ok, good spot - will fix tmrw

murggel commented:
I tried a little with the layout:
Removing the padding from the outer LinearLayout and adding the 10dp margin to all borders leads to the "modified new dialog"-appearance, which is closer to the original dialog.

What do you think?

| current new dialog | modified new dialog |
| -- | -- |
|![image](https://github.com/user-attachments/assets/7649a15a-3659-4e01-b023-ae546148517d) |![image](https://github.com/user-attachments/assets/fa15f9fd-bece-4bb3-8f2d-9e230acc834a)|

![image](https://github.com/user-attachments/assets/fa9bd1f6-053a-4556-962a-97152046ccc3)

murggel commented:
> Minor UI-issue: It would be nice, if the ° and ' signs are in the input fields of the corresponding format-fields ![image](https://private-user-images.githubusercontent.com/70113341/435277195-009fb32e-d280-4279-a624-afb17633740c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNDY1OTcsIm5iZiI6MTc0NTM0NjI5NywicGF0aCI6Ii83MDExMzM0MS80MzUyNzcxOTUtMDA5ZmIzMmUtZDI4MC00Mjc5LWE2MjQtYWZiMTc2MzM3NDBjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIyVDE4MjQ1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY3NjE2MGNiN2QzNzZmNzM2YTYyMzVhNmFjZjlhM2RiNDdkZmE1ZDkyODgwODcyNDQ5ZDUxYjIyMDIyMTY4YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KJxIWjOSt45Z8WBNqR7oLhQBKW1-jg8y_-oRscClobc)

This was not solved with reusing the original layout?

ReggieDoodle commented:
> This was not solved with reusing the original layout?

No, the layout I copied was for the plain text fields where you asked for the cancel button.  

The degree min and second signs are in code but I need to add the associated separator fields to the layout to allow that.

moving-bits commented:
Thanks @ReggieDoodle for your hard work!

ReggieDoodle commented:
Tks. 

Pleased to be able to contribute.

I will tie up the loose ends today and make a new PR.

I've already converted all usages except the waypoint page to the new dialog (search card, user defined cache and GK TBs).  Do you want these in there too ?
